node/src/components/consensus/protocols/highway/synchronizer.rs

```diff
@@ -0,0 +1,258 @@
+        highway_core::highway::{Dependency, Highway, PreValidatedVertex, Vertex},
+        traits::{Context, NodeIdT},
+    },
+    types::{TimeDiff, Timestamp},
+};
+
+use super::{ProtocolOutcomes, ACTION_ID_VERTEX};
+
+/// An incoming vertex that we haven't added to the protocol state yet.
+#[derive(DataSize, Debug)]
+pub(crate) struct PendingVertex<I, C>
+where
+    C: Context,
+{
+    /// The peer who sent it to us.
+    sender: I,
+    /// The pre-validated vertex.
+    pvv: PreValidatedVertex<C>,
+    /// The time when we received it.
+    time_received: Timestamp,
+}
+
+impl<I, C: Context> PendingVertex<I, C> {
+    /// Returns a new pending vertex with the current timestamp.
+    pub(crate) fn new(sender: I, pvv: PreValidatedVertex<C>) -> Self {
+        Self {
+            sender,
+            pvv,
+            time_received: Timestamp::now(),
+        }
+    }
+
+    pub(crate) fn sender(&self) -> &I {
+        &self.sender
+    }
+
+    /// Returns the vertex waiting to be added.
+    pub(crate) fn vertex(&self) -> &Vertex<C> {
+        self.pvv.inner()
+    }
+
+    pub(crate) fn pvv(&self) -> &PreValidatedVertex<C> {
+        &self.pvv
+    }
+
+    /// Returns `true` if this vertex was received longer than `timeout` ago.
+    fn expired(&self, timeout: TimeDiff) -> bool {
+        self.time_received + timeout < Timestamp::now()
+    }
+}
+
+impl<I, C: Context> Into<PreValidatedVertex<C>> for PendingVertex<I, C> {
+    fn into(self) -> PreValidatedVertex<C> {
+        self.pvv
+    }
+}
+
+#[derive(DataSize, Debug)]
+pub(crate) struct Synchronizer<I, C>
+where
+    C: Context,
+{
+    /// Incoming vertices we can't add yet because they are still missing a dependency.
+    vertex_deps: BTreeMap<Dependency<C>, Vec<PendingVertex<I, C>>>,
+    /// The vertices that are scheduled to be processed at a later time.  The keys of this
+    /// `BTreeMap` are timestamps when the corresponding vector of vertices will be added.
+    vertices_to_be_added_later: BTreeMap<Timestamp, Vec<PendingVertex<I, C>>>,
+    /// Vertices that might be ready to add to the protocol state: We are not currently waiting for
+    /// a requested dependency.
+    vertices_to_be_added: Vec<PendingVertex<I, C>>,
+    /// The duration for which incoming vertices with missing dependencies are kept in a queue.
+    pending_vertex_timeout: TimeDiff,
+}
+
+impl<I: NodeIdT, C: Context + 'static> Synchronizer<I, C> {
+    pub(crate) fn new(pending_vertex_timeout: TimeDiff) -> Self {
+        Synchronizer {
+            vertex_deps: BTreeMap::new(),
+            vertices_to_be_added_later: BTreeMap::new(),
+            vertices_to_be_added: Vec::new(),
+            pending_vertex_timeout,
+        }
+    }
+
+    /// Removes expired pending vertices from the queues, and schedules the next purge.
+    pub(crate) fn purge_vertices(&mut self) {
+        let timeout = self.pending_vertex_timeout;
+        self.vertices_to_be_added.retain(|pv| !pv.expired(timeout));
```

Comment:
Do we need to iterate through this collection? The comment says:
```
Vertices that might be ready to add to the protocol state: We are not currently waiting for a requested dependency.
```
which means that we're not blocked by a spammy peer and we have all the information necessary to add it.

Reply:
The vertices in that queue could still be missing dependencies: It's just that the last dependency we knew of was satisfied, so we'll try again when the event fires. That might reveal more missing dependencies and add them back to `vertex_deps`.

Reply:
Ah, ok. I haven't really checked how it gets populated.